sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,207 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include "sw/device/lib/dif/dif_padctrl.h"
+
+// Generated.
+// Note that if the the number of DIO or MIO pads is such that all attribute can
+// fit in one register the PADCTRL_DIO_PADS0_REG_OFFSET and
+// PADCTRL_MIO_PADS0_REG_OFFSET #defines from padctrl_regs.h will no longer
+// contain '0' and become PADCTRL_DIO_PADS_REG_OFFSET and
+// PADCTRL_MIO_PADS_REG_OFFSET respectively which will require altering their
+// names in the source below.
+#include "padctrl_regs.h"
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+
+static const int kPadctrlRegisterWidthInBytes = (PADCTRL_PARAM_REG_WIDTH / 8);
+
+// This is the value written to a whole attribute register, it is not a per-pad
+// value so desired default attribute setting must be replicated per attribute
+// field in the attribute register.
+static const uint32_t kPadctrlAttrDefault = 0;
+static const uint32_t kPadctrlAttrFieldMask = PADCTRL_MIO_PADS0_ATTR0_MASK;
+
+/**
+ * Checks if an attribute field contains conflicting attributes
+ */
+static bool conflicting_attributes(uint32_t attrs) {
+  // Cannot be both pull up and pull down
+  if ((attrs & (1 << kDifPadctrlAttrPullDown)) &&
+      (attrs & (1 << kDifPadctrlAttrPullUp))) {
+    return true;
+  }
+
+  // Cannot combine pull down with open drain
+  if ((attrs & (1 << kDifPadctrlAttrPullDown)) &&
+      (attrs & (1 << kDifPadctrlAttrOpenDrain))) {
+    return true;
+  }
+
+  return false;
+}
+
+dif_padctrl_init_result_t dif_padctrl_init(mmio_region_t base_addr,
+                                           dif_padctrl_t *padctrl) {
+  if (padctrl == NULL) {
+    return kDifPadctrlInitBadArg;
+  }
+
+  bool padctrl_unlocked = mmio_region_get_bit32(
+      base_addr, PADCTRL_REGEN_REG_OFFSET, PADCTRL_REGEN_WEN);
+
+  if (!padctrl_unlocked) {
+    return kDifPadctrlInitLocked;
+  }
+
+  padctrl->base_addr = base_addr;
+
+  // Reset all DIO pad attributes to the default value
+  for (int i = 0; i < PADCTRL_DIO_PADS_MULTIREG_COUNT; ++i) {
+    ptrdiff_t dio_pads_register_offset =
+        PADCTRL_DIO_PADS0_REG_OFFSET + (i * kPadctrlRegisterWidthInBytes);
+    mmio_region_write32(padctrl->base_addr, dio_pads_register_offset,
+                        kPadctrlAttrDefault);
+  }
+
+  // Reset all MIO pad attributes to the default value
+  for (int i = 0; i < PADCTRL_MIO_PADS_MULTIREG_COUNT; ++i) {
+    ptrdiff_t mio_pads_register_offset =
+        PADCTRL_MIO_PADS0_REG_OFFSET + (i * kPadctrlRegisterWidthInBytes);
+    mmio_region_write32(padctrl->base_addr, mio_pads_register_offset,
+                        kPadctrlAttrDefault);
+  }
```

Comment:
Garret and I discussed the reset behaviour today when looking at pinmux, and our feeling is that this DIF should not reset the registers back to known good values, it should leave them as it got them.